metadata-ingestion/src/datahub/ingestion/source/bigquery_v2/bigquery.py

```diff
@@ -1283,6 +1222,64 @@ def get_tables_for_dataset(
 
         return bigquery_tables
 
+    def get_core_table_details(
+        self, conn: bigquery.Client, dataset_name: str, project_id: str
+    ) -> Dict[str, TableListItem]:
+        table_items: Dict[str, TableListItem] = {}
+        # Dict to store sharded table and the last seen max shard id
+        sharded_tables: Dict[str, TableListItem] = defaultdict()
+
+        for table in conn.list_tables(f"{project_id}.{dataset_name}"):
+            table_identifier = BigqueryTableIdentifier(
+                project_id=project_id,
+                dataset=dataset_name,
+                table=table.table_id,
+            )
+
+            _, shard = BigqueryTableIdentifier.get_table_and_shard(
+                table_identifier.raw_table_name()
+            )
+            table_name = table_identifier.get_table_name().split(".")[-1]
+
+            # Sharded tables look like: table_20220120
+            # For sharded tables we only process the latest shard and ignore the rest
+            # to find the latest shard we iterate over the list of tables and store the maximum shard id
+            # We only has one special case where the table name is a date `20220110`
+            # in this case we merge all these tables under dataset name as table name.
+            # For example some_dataset.20220110 will be turned to some_dataset.some_dataset
+            # It seems like there are some bigquery user who uses this non-standard way of sharding the tables.
+            if shard:
+                if not sharded_tables.get(table_name):
+                    # When table is only a shard we use dataset_name as table_name
+                    sharded_tables[table_name] = table
+                    continue
+
+                stored_table_identifier = BigqueryTableIdentifier(
+                    project_id=project_id,
+                    dataset=dataset_name,
+                    table=sharded_tables[table_name].table_id,
+                )
+                _, stored_shard = BigqueryTableIdentifier.get_table_and_shard(
+                    stored_table_identifier.raw_table_name()
+                )
+                # When table is none, we use dataset_name as table_name
+                assert stored_shard
```

Comment:
Do we have exception handling for this `assert`? What does this statement do to the execution flow of ingestion?

Reply:
This should not happen and it is for mypy

Reply:
I see. I guess if we wanted to be more type safe we could make `sharded_tables` store mapping `table_name -> (table, shard)` but doesn't seem necessary

Reply:
Perhaps we should wrap with a try catch and log an error just in case we later make a code change that adds a non-sharded table to this dict